go/libraries/doltcore/sqle/dtables/unscoped_diff_table.go

```diff
@@ -189,39 +202,158 @@ func (itr *UnscopedDiffTableItr) calculateChangedTables(commit *doltdb.Commit) (
-		// Use toName by default for the table name by default, but if it's
-		// nil (e.g. when dropping a table), check to see if fromName is available.
-		tableName := delta.ToName
-		if len(tableName) == 0 {
-			if len(delta.FromName) > 0 {
-				tableName = delta.FromName
-			}
+		change, err := itr.processTableDelta(delta)
+		if err != nil {
+			return nil, err
 		}
 
-		if len(tableName) > 0 {
-			tablesMap[tableName] = true
-		}
+		tableChanges = append(tableChanges, *change)
 	}
 
 	// Not all commits mutate tables (e.g. empty commits)
-	if len(tablesMap) == 0 {
+	if len(tableChanges) == 0 {
 		return nil, nil
 	}
 
-	tables := make([]string, len(tablesMap))
-	i := 0
-	for key := range tablesMap {
-		tables[i] = key
-		i++
+	sort.Slice(tableChanges, func(i, j int) bool {
+		return tableChanges[i].tableName < tableChanges[j].tableName
+	})
+
+	return tableChanges, nil
+}
+
+// processTableDelta processes the specified TableDelta to determine what kind of change it was (i.e. table drop,
+// table rename, table create, or data update) and returns a tableChange struct representing the change.
+func (itr *UnscopedDiffTableItr) processTableDelta(delta diff.TableDelta) (*tableChange, error) {
+	// Dropping a table is always a schema change, and also a data change if the table contained data
+	if itr.isTableDropChange(delta) {
+		isEmpty, err := itr.isTableDataEmpty(delta.FromTable)
+		if err != nil {
+			return nil, err
+		}
+
+		return &tableChange{
+			tableName:    delta.FromName,
+			dataChange:   !isEmpty,
+			schemaChange: true,
+		}, nil
+	}
+
+	// Renaming a table is always a schema change, and also a data change if the table data differs
+	if itr.isRenameChange(delta) {
+		dataChanged, err := itr.isTableDataDifferent(delta)
+		if err != nil {
+			return nil, err
+		}
+
+		return &tableChange{
+			tableName:    delta.ToName,
+			dataChange:   dataChanged,
+			schemaChange: true,
+		}, nil
 	}
-	sort.Strings(tables)
 
-	return tables, nil
+	// Creating a table is always a schema change, and also a data change if data was inserted
+	if itr.isTableCreateChange(delta) {
+		isEmpty, err := itr.isTableDataEmpty(delta.ToTable)
+		if err != nil {
+			return nil, err
+		}
+
+		return &tableChange{
+			tableName:    delta.ToName,
+			dataChange:   !isEmpty,
+			schemaChange: true,
+		}, nil
+	}
+
+	dataChanged, err := itr.isTableDataDifferent(delta)
+	if err != nil {
+		return nil, err
+	}
+
+	schemaChanged, err := itr.isTableSchemaDifferent(delta)
+	if err != nil {
+		return nil, err
+	}
+
+	return &tableChange{
+		tableName:    delta.ToName,
+		dataChange:   dataChanged,
+		schemaChange: schemaChanged,
+	}, nil
 }
 
 // Close closes the iterator.
 func (itr *UnscopedDiffTableItr) Close(*sql.Context) error {
 	return nil
 }
+
+// isTableDataEmpty return true if the table does not contain any data
+func (itr *UnscopedDiffTableItr) isTableDataEmpty(table *doltdb.Table) (bool, error) {
+	rowData, err := table.GetNomsRowData(itr.ctx)
```

Comment:
Probably want to use GetRowData here (storage-agnostic version of this) if possible